### PR TITLE
Native for loop's head has destructuring declaration which must have an initializer.

### PR DESCRIPTION
**Pull request by akroshg**
Native for loop's head has destructuring declaration which must have an initializer.

Native for loop with destructuring does not enforce that it must have an initializer. For that purpose the ParseVariableDeclaration takes another param which is actually deducing whether native for loop is okay or not. We do this kind of checking as we don't know whether current loop turned out to be native loop or for..in/of loop.
If the current syntax turned out to be native for loop and during declaration we deduced that native for loop will not be okay then we throw error.
While playing with different test cases  I found that TrackAssignment assert requires adjustment as in the case of for..in loop we rescan the loop head again for validating destructuring grammar. Due that assert is not valid as the symbol span require re-adjustment.
Added few tests to cover more cases.

### CODE REVIEW

**Comment by akroshg** [comment]
Fixes #225 

@tcare , @ianwjhalliday  could you review pls?

File: lib/Parser/parse.cpp [hunk 1]
```diff
@@ -7830,7 +7830,8 @@ ParseNodePtr Parser::ParseVariableDeclaration(
     BOOL singleDefOnly/* = FALSE*/,
     BOOL allowInit/* = TRUE*/,
     BOOL isTopVarParse/* = TRUE*/,
-    BOOL isFor/* = FALSE*/)
+    BOOL isFor/* = FALSE*/,
+    BOOL* nativeForOk /*= nullptr*/)
```
Comment (akroshg) [inline lib/Parser/parse.cpp]: Going for BOOL just for consistency in the current method.

File: lib/Parser/parse.cpp [hunk 2]
```diff
@@ -7708,7 +7708,7 @@ void PidRefStack::TrackAssignment(charcount_t ichMin, charcount_t ichLim)
         {
             return;
         }
-        Assert(this->GetIchLim() >= ichLim);
+        Assert(ichMin <= this->GetIchMin() && this->GetIchLim() <= ichLim)
```
Comment (akroshg) [inline lib/Parser/parse.cpp]: Just realized semicolon is missing above statement. will fix this in next rev.

Comment (ianwjhalliday) [inline lib/Parser/parse.cpp]: Oh.  `#define AssertMsg` (in Assertions.h) should be wrapped in `do { ... } while (false)` to force a semicolon to come after the usage of the macro.

Comment (akroshg) [inline lib/Parser/parse.cpp]: are you looking at lib\common\core\Assertions.h? So debug build is fine without semicolon, but non-dbg will have problem.

Comment (tcare) [inline lib/Parser/parse.cpp]: +1 to do while to force semicolon.

Comment (ianwjhalliday) [inline lib/Parser/parse.cpp]: Development is done with debug builds so makes sense to make it an error in debug builds.  The do while trick forces it to be a statement that requires a semicolon.

Comment (akroshg) [inline lib/Parser/parse.cpp]: Yeah, I have done that do..while change locally - will do different PR for that.

**Comment by tcare** [comment]
Looks good

**Comment by akroshg** [comment]
Thanks @tcare
